File: April 2020/30 - Check If a String Is a Valid Sequence form Root to Leaves Path in a Binary Tree/Solution.py
class Solution:
    def isValidSequence(self, root: TreeNode, arr: List[int]) -> bool:
        n = len(arr)
        def dfs(node, idx):
            if not node or node.val != arr[idx]:
                return False
            if idx == n - 1:
                return not (node.left or node.right)
            return dfs(node.right, idx + 1) or dfs(node.left, idx + 1)
        return dfs(root, 0)

File: April 2020/30 - Check If a String Is a Valid Sequence form Root to Leaves Path in a Binary Tree/Solution.java
class Solution {
    public boolean isValidSequence(TreeNode root, int[] arr) {
        return dfs(root, arr, 0);
    }
    
    private boolean dfs(TreeNode node, int[] arr, int idx) {
        if(node == null || node.val != arr[idx])
            return false;
        if(idx == arr.length - 1)
            return node.left == null && node.right == null;
        return dfs(node.right, arr, idx + 1) || dfs(node.left, arr, idx + 1);
    }
}